Zis aes - Fix HMAC Calculation

remogloor opened the pull request:
Hi 

I took a look at your draft for ZipInputStream AES support. The problem with the HMAC calculation is that your implementation always decrypted and calculated HMAC for the entire input buffer. But actually it should only decrypt and calculate the HMAC up to the compressed size of the current entry. This pull request fixes that problem and all the Unit Tests pass.

You might want to look into that implementation again as I don't know from your changes if there is anything else that needs to be changed.

Best regards
Remo

------
I certify that I own, and have sufficient rights to contribute, all source code and related material intended to be compiled or integrated with the source code for the SharpZipLib open source product (the "Contribution"). My Contribution is licensed under the MIT License.

Numpsy committed:
Add support for AES encryption

Numpsy committed:
Merge PR #432: Throw ArgumentNullException in BZip2

Change the BZip2 Compress/Decompress functions to throw ArgumentNullxception rather than Exception when null stream parameters are provided

Numpsy committed:
Merge PR#469: Add test for writing using a zero byte buffer

Numpsy committed:
Merge PR #473: Add a Nuget badge to readme.md

bastianeicher committed:
Merge PR #463: Improve support for Unix timestamps in ZIP archives

* Store ZipEntry.DateTime in dedicated backing field  
This allows reading values with a higher resolution than DOS time (2 second accuracy).

* Fix getting unix modification time in ZIP files
InfoZIP actually does respect a file's modification time, even if the access time and/or creation time are not set.

* Remove backing field for ZipEntry.DosTime, convert to and from ZipEntry.DateTime instead
This prevents the two values from becoming potentially inconsistent.

Numpsy committed:
Merge PR #468: Add test for adding empty folders to archives using FastZip

Helios-vmg committed:
Merge PR #467: Allow seeking a PartialInputStream to the very end

Numpsy committed:
Merge PR #466: Improve the ZipFileStoreAesPartialRead test to test multiple block sizes

Helios-vmg committed:
Merge PR #465: Fixed bug in ZipAESStream.ReadBufferedData

copyCount was being computed but not passed to Array.Copy(), which caused an
exception to be thrown when enough bytes were read from an encrypted stream.

Numpsy committed:
Merge PR #453: Fix the 7-zip interop tests in the .Net 4.6 test build

Numpsy committed:
Merge PR #461: Fix Exception doc comments

* fix doc comment in ZipInputStream.BodyRead
* fix doc comment in ZipOutputStream.SetComment

Numpsy committed:
Merge PR #448: Fix unit test assert argument order

itn3000 committed:
Merge PR #364: Add nameEncoding parameter to Tar entries

* add encoding parameter to creating tar entry

default is same as current master behavior(omit upper byte)

* add encoding tests(cp932) and add doc comment(#364)

* add header bytes test and mark obsolete methods without encodings(#364)

but IEntryFactory does not considering name encoding.

* forget to mark as obsoleting(#364)

* add doc comment for name encoding parameter (#364)

Numpsy committed:
Merge PR #472: Allow ZipFile to accept empty strings as passwords when decrypting AES entries

* Add unit test for reading an AES encrypted entry with an empty password
* Allow ZipFile to accept empty strings as passwords when decrypting AES entries

Numpsy committed:
Merge PR #390: Override Ensure GZipOutputStream headers are written before flush

* Add unit tests to repro #382
* Add an override of Flush() to GZipOutputStream to ensure the headers is writen before flushing

Numpsy committed:
Merge PR #435: Add unit test for ZipFile.Add(string fileName, string entryName)

Numpsy committed:
Merge PR #433: Restore directory timestamps when extracting with FastZip

* When extracting folders with FastZip, reset the last modified time if the RestoreDateTimeOnExtract option is enabled
* Add unit test for restoring directory timestamps when extracing with fastzip and RestoreDateTimeOnExtract is true

Numpsy committed:
PR #450: Fix CA1200 code analyzer warnings

Numpsy committed:
PR #445: Make InvalidHeaderException serializable

Numpsy committed:
PR #422: Change ZipOutputStream.PutNextEntry to explicity validate the requested compression method

Numpsy committed:
PR #479: Streamline and update VB sample projects

Numpsy committed:
PR #477: Fix spelling errors in comments

Numpsy committed:
PR #476: Remove duplicated words in comments

Co-authored-by: nils måsén <nils@piksel.se>

christophwille committed:
SharpZipLib is now using GH Discussions and no longer gitter.im

christophwille committed:
Remove state of the union from 2017

Numpsy committed:
PR #498: Use string.Trim to trim strings (#498)

Numpsy committed:
PR #494: Use the Range to test different compression levels in InflaterDeflaterTestSuite

Numpsy committed:
PR #489: Remove duplicate ICSharpCode.SharpZipLib.snk

Numpsy committed:
#488: Add [MemoryDiagnoser] to the zip input/output stream benchmark classes

Numpsy committed:
PR #455: Add FastZip.CreateZip with a leaveOpen parameter

* Add a variant of FastZip.CreateZip with a leaveOpen parameter to control output stream disposal
* Add unit test for FastZip.CreateZip leaving the stream open or disposed as required

Numpsy committed:
PR #458: Dispose of entry streams returned by ZipFile.GetInputStream

Numpsy committed:
PR #483: Suppress CA1707 warnings from LzwConstants

Numpsy committed:
PR #460: Account for AES overhead in compressed entry size (#460)

* Unit test for writing encrypted 0 byte files to ZipOutputStream
* In ZipOutputStream.PutNextEntry, account for AES overhead when calculating compressed entry size

Commit:
PR #402 Only convert entry.Name once when accessing updateIndex

* don't convert entry.Name twice when accessing updateIndex

* Update src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs

Accepted suggestion

Co-Authored-By: nils måsén <nils@piksel.se>

* Update src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs

Co-Authored-By: nils måsén <nils@piksel.se>

* Update src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs

Co-Authored-By: nils måsén <nils@piksel.se>

* Update src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs

Co-Authored-By: nils måsén <nils@piksel.se>

Co-authored-by: nils måsén <nils@piksel.se>

piksel committed:
PR #502: Fix tests and ZipEntry DateTime Kind

* Normalize DateTime kind in ZipEntry

* Use Unspecifed DateTime Kind for entries

* Fix tests

Numpsy committed:
PR #482: Add variants of FastZip.CreateZip taking IScanFilter instead of strings

Numpsy committed:
PR #497: Transform new entry names using an INameTranform in ZipOutputStream

Numpsy committed:
PR #353: Fix ZipFile.TestLocalHeader CompressionMethod resolving for AES entries

* Change ZipFile.TestLocalHeader to check CompressionMethodForHeader rather than CompressionMethod.

* Test ZipFile.TestArchive with an AES encrypted archive.

Numpsy committed:
PR #380: Add support for AES encryption in FastZip.CreateZip

Numpsy committed:
PR #504: Fix warning about missing doc comment in FastZip.CreateZip

piksel committed:
PR #505: Expect ZipEntry clean name test to be positive

piksel committed:
Use github workflows for CI (#419)

* Add Github Actions workflows

* Use VS 2019 in AppVeyor CI

Stevie-O committed:
PR #201: Raise ProcessDirectory event for FastZip extract

FastZip's ExtractZip method did not raise the ProcessDirectory event when
creating a directory.  There was no way to determine or control when
ExtractZip decided to create a new directory.

NOTE: The logic to raise the event only triggers when the target directory
doesn't already exist in the destination.

Numpsy committed:
PR #333: Handle unsupported compression methods in ZipInputStream better

* Change ZipInputStream to use its own IsEntryCompressionMethodSupported function rather than the one in ZipEntry.

* Unit test for ZipInputStream.CanDecompressEntry being false for AES encrypted entries.

* Fix comment typo

Co-authored-by: nils måsén <nils@piksel.se>

piksel committed:
PR #346: Add a Security Policy

piksel committed:
Fix master on-push workflow

piksel committed:
Make healthy sourcelink nupkgs

Numpsy committed:
PR #451: Minimize and update sample app package dependencies

Numpsy committed:
PR #452: Ensure crypto streams are disposed in ZipFile.GetOutputStream

Numpsy committed:
PR #503: Consider AES overhead when testing encrypted folder entries

* Extend the ZipFile.TestEncryptedDirectoryEntry test to test AES encrypted directories

* Consider AES overhead when testing encrypted folder entries

piksel committed:
PR #519: Restrict path traversal on TarArchive extraction

piksel committed:
Update ICSharpCode.SharpZipLib.csproj

piksel committed:
Add Codacy to PR workflow

piksel committed:
Add Codacy to Push workflow

piksel committed:
Fix typo in Push workflow

piksel committed:
PR #523: Fix Codacy workflows

* Fix typo and set max allowed to 9999
* Add codacy config
* Update on-push.yml
* Add project token to Push WF
* Add project token to PR WF
* Add Codacy upload to push WF
* Add Codacy upload to PR WF

piksel committed:
PR #522: Use PackageIcon instead of PackageIconUrl in csproj

* Update ICSharpCode.SharpZipLib.csproj
* Add nuget icon
* Fix assets path

Numpsy committed:
PR #355: Wire up BZip2 compression support for Zip files

Numpsy committed:
PR #525: Update the sample projects to use the v1.3 release

Numpsy committed:
PR #511: Move the 7zip helper functions into their own class for easier reuse

Numpsy committed:
PR #509: Make PutNextEntry throw if the entry is AES and no password has been set

refs #507

piksel committed:
#PR 529: Add release workflow

piksel committed:
PR #530: Fix Codacy Workflows

Updates the Codacy Action to v1.1.0

piksel committed:
PR #539: Use securely generated random temporary file names

- Use GetRandomFileName() instead of  GetTempFileName()
- Unify random file name creation into PathUtils
- Do not create new random files before use
- Rename  WindowsPathUtils to PathUtils and make a proper static class

piksel committed:
PR #538: Use RNGCryptoServiceProvider for crypto headers

- Use RNGCryptoServiceProvider for crypto headers
- Dispose RNG service after use in ZipCrypto

nsgomez committed:
PR #516: Improve CRC32 performance

piksel committed:
PR #475: Fix updating zips with descriptor entries

- Unit tests for using ZipFile to update file entries that have descriptors
- Fix size calculation in GetDescriptorSize
- Handle optional descriptor signature when updating

Numpsy committed:
PR #475: Fix updating zips with descriptor entries

Dummy commit since Github didn't recognize the merge

Co-authored-by: nils måsén <nils@piksel.se>

piksel committed:
PR #541: Update csproj for new release

piksel committed:
CI: run master build on release

piksel committed:
#542: Build and publish documentation on release

Numpsy commented:
It's probably better to leave any discussion to the SharpZipLib repo, where more people will see it

piksel committed:
PR #553: Remove broken codacy integration

0xced committed:
PR #549: Add .NET Standard 2.1 target framework

Also fix the .NET Standard 2.0 target framework moniker: netstandard2.0 instead of netstandard2, see https://docs.microsoft.com/en-us/dotnet/standard/frameworks#supported-target-frameworks

piksel committed:
PR #554: Skip CRC calculation for AES zip entries

Numpsy committed:
PR #546: Make pure private functions static

Numpsy committed:
PR #533: Convert the C# sample projects to PackageReference format

* Convert the C# sample projects to PackageReference format
* Update example package ref to v1.3.1

Co-authored-by: nils måsén <nils@piksel.se>

Numpsy committed:
PR #457: add basic async unit tests for the inflator/deflator streams

* add some basic async unit tests for the deflator streams

Co-authored-by: nils måsén <nils@piksel.se>

piksel committed:
chore(ci): fix windows ci env restore

See https://github.com/actions/virtual-environments/issues/1090

piksel committed:
chore(ci): remove config from matrix

Numpsy committed:
PR #510: Build the test bootstrapper app as netcoreapp3.1 instead of netcoreapp2.0

Numpsy committed:
PR #577: Throw ZipException in ZipAESStream instead of generic Exception

Numpsy committed:
PR #578: Fix typos in the StreamDecodingException doc comments

piksel committed:
PR #517: Throw exception on Store+Descriptor entries

* fix(zip): indicate that store/desc. entries cant be extracted
* style(zip): simplify and fix spelling
* fix(zip): check decompress support after specific exception handling
* test: use random bad path instead of possible network drive
* docs(zip): remove explicit mentions of methods in CompressionMethod
* fix(zip): remove superfluous null check

Numpsy committed:
PR #588: Add a simple async read test for ZipFile

Numpsy committed:
PR #586: Convert VB sample projects to PackageReference format

piksel committed:
PR #583: Restore entry times on FastZip extract

* FastZip - fix TimeSetting control when extracting Zip entries
* Add tests for FastZip file times

Co-authored-by: va4es2 <va4es2@users.noreply.github.com>

ref #555 #566

Numpsy committed:
PR #575: Replace uses of new T[0] with Array.Empty<T>

* Add the 'EmptyRefs' helper, to support Array.Empty on .NET 4.5 (ref: #501)
* Replace uses of 'new T[0]' with 'Empty.Array<T>'

piksel committed:
Remove custom nuget config

Numpsy committed:
PR #594: Remove the local nuget.config files from the test projects

piksel committed:
PR #599: Use net46 as CI target framework for windows tests

* use net46 for windows testing on push
* use net46 for windows testing for PRs

piksel committed:
Split build and pack in CI on push

piksel committed:
PR #601: pass tests that are within time tolerance

piksel committed:
PR #602: Update test/coverage packages and push to codecov

* update test/coverage packages and push to codecov
* update bootstrapper packages
* upload coverage on push

piksel committed:
PR #603: pass CreateZip tests that are within time tolerance

piksel committed:
chore(ci): only export coverage for release

piksel committed:
test: retry zipcrypto password tests once (#614)

This is due to the password check only comparing a single byte (as per the spec), which have a 1/255 chance of actually matching regardless of password.

Numpsy committed:
PR #605: Suppress CA1707 warnings in the Constants classes

piksel committed:
PR #593: Simplify DropPathRoot and fix out of bounds issue

* Simplify DropPathRoot and fix out of bounds issue

* prevent throwing in DropPathRoot and add tests

* make path tests platform-specific

testing for UNC paths on *nix does not really make sense

* handle .NET < 4.6.2 special cases

piksel committed:
PR 351: Add support for Filename field in GZip

Numpsy committed:
PR #616: Change the build status badge to reference github actions

konrad-kruczynski committed:
PR #611: Bzip input stream simple vectorization

* Added benchmark for BZip2 decompression.
* Updated benchmarks to be run also on .NET Core 3.1.
* Simple automatic vectorization of the rotation loop.
* Added comment describing vectorization.

piksel committed:
PR #621: unify PR/push CI build actions

* ci: unify PR/push CI build actions
* ci: build packages using .net 5 for healthy nupkgs
* ci: simplify action syntax, use correct build flags

Numpsy committed:
PR #587: Remove supported method checks from ZipEntry.CompressionMethod setter

Numpsy committed:
#579: Implement very simple ReadAsync in ZipAESStream

* add an extra unit test for doing an async read on a ZipFile InputStream
* Implement ReadAsync in ZipAESStream, extra simple version

piksel committed:
fix CI badge in README

piksel committed:
PR #622: security fixes for v1.3.2

* test: alter LimitExtractPath to check for file/dir collision
* fix: disallow traversal when file and base dir share name
* fix: use random file name for writing asciitrans tar entries
* fix: add dir separator to base dir if missing

piksel committed:
update csproj for v1.3.2 release

Numpsy committed:
PR #627: fix the expected/actual value ordering in unit tests

Numpsy committed:
PR #626: make a couple of private functions static

Numpsy committed:
PR #625: Make BZip2Constants static instead of sealed

piksel committed:
PR #628: Limit code coverage to windows CI

Numpsy committed:
PR #604: Move the Password property from DeflaterOutputStream into ZipOutputStream

* Move the Password property from DeflaterOutputStream into ZipOutputStream
* Move cryptoTransform setup machinery from DeflatorOutputStream to ZipOutputStream
* remove duplicate setup of _aesRnd

Numpsy committed:
PR #634: add an async version of the WriteZipOutputStream benchmark

piksel committed:
PR #636: Fix unstable tests and switch to dotcover

* test: use static random seed in tests
  this prevents the code coverage from varying depending on how
  well the random data compresses
* ci: use dotcover for code coverage

piksel committed:
fix(tar): create translated files in temp (#645)

ForNeVeR committed:
docs(zip): fix ZipStrings typo (#648)

piksel committed:
ci: add codeql analysis

modio-jackson committed:
fix(bzip2): use explicit feature defs for vectorized memory move (#635)

* Fixed mismatched framework directives for vectorized memory move

Co-authored-by: nils måsén <nils@piksel.se>

piksel committed:
update security policy

the security incorrectly indicated that minor versions would receive backports of security fixes. 
as long as the API is fully backwards-compatible (as indicated by semver), there should be no need to release additional patches for older versions.

piksel committed:
test: add tests for tar path traversal

piksel committed:
fix: specialized tar extract traversal

piksel committed:
update csproj for v1.3.3 release

piksel committed:
ci: add manual dispatch and logging

piksel committed:
ci: fix yaml typo

piksel committed:
ci: include non-anno tags in version

piksel committed:
chore: add release notes tooling (#665)

ueli-werner committed:
docs(zip): fix slash names in CleanName comment (#668)

piksel committed:
fix(tar): permit slashed output dir (#666)

piksel committed:
test: repeatability and refactoring (#671)

Numpsy committed:
fix(zip): use ushort for ITaggedData.TagID (#669)

Numpsy committed:
chore: remove the test bootstrapper (#670)

piksel committed:
feat(zip): ZipOutputStream async support (#574)

Co-authored-by: Cédric Luthi <cedric.luthi@gmail.com>

piksel committed:
feat(zip): better string encoding handling (#592)

This replaces the global static ZipStrings singleton with instances of StringCodec, which will:

- Remove encoding configuration from a shared global state
- Allow for different defaults for input and output
- Explicitly override the encodings used for ZipCrypto and zip archive comments (the one in the Central Directory, not the individual entry comments).
- Use "Unicode" for new entries (unless overriden)
- Make it much more clear (hopefully) how and why different encodings are used.

Temtaime committed:
feat(zip): enable ZipOuputStream to write precompressed files (#683)

Co-authored-by: nils måsén <nils@piksel.se>

dependabot[bot] committed:
Bump SharpZipLib in /samples/ICSharpCode.SharpZipLib.Samples/cs/sz (#716)

Bumps [SharpZipLib](https://github.com/icsharpcode/SharpZipLib) from 1.3.1 to 1.3.3.
- [Release notes](https://github.com/icsharpcode/SharpZipLib/releases)
- [Changelog](https://github.com/icsharpcode/SharpZipLib/blob/master/docs/Changes.txt)
- [Commits](https://github.com/icsharpcode/SharpZipLib/compare/v1.3.1...v1.3.3)

---
updated-dependencies:
- dependency-name: SharpZipLib
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump SharpZipLib (#715)

Bumps [SharpZipLib](https://github.com/icsharpcode/SharpZipLib) from 1.3.1 to 1.3.3.
- [Release notes](https://github.com/icsharpcode/SharpZipLib/releases)
- [Changelog](https://github.com/icsharpcode/SharpZipLib/blob/master/docs/Changes.txt)
- [Commits](https://github.com/icsharpcode/SharpZipLib/compare/v1.3.1...v1.3.3)

---
updated-dependencies:
- dependency-name: SharpZipLib
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump SharpZipLib in /samples/ICSharpCode.SharpZipLib.Samples/cs/zf (#711)

Bumps [SharpZipLib](https://github.com/icsharpcode/SharpZipLib) from 1.3.1 to 1.3.3.
- [Release notes](https://github.com/icsharpcode/SharpZipLib/releases)
- [Changelog](https://github.com/icsharpcode/SharpZipLib/blob/master/docs/Changes.txt)
- [Commits](https://github.com/icsharpcode/SharpZipLib/compare/v1.3.1...v1.3.3)

---
updated-dependencies:
- dependency-name: SharpZipLib
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

piksel committed:
fix(samples): bump lib versions to v1.3.3 (#720)

piksel committed:
ci: use server 2019 for win builds

piksel committed:
ci: use server 2019 for win builds (#727)

piksel committed:
fix(zip): dont fail test on 0 sizes and descriptor (#736)

piksel committed:
add openupm package badge

piksel committed:
feat(zip): make it possible to skip header tests (#689)

piksel committed:
fix(zip): 0 in zip64 local sizes using descriptors (#750)

nathan-c committed:
feat(tar): support for async streams (#746)

piksel committed:
fix(zip): skip reading position for non-seekable async streams (#754)

yihezkel committed:
fix(zip): explicitly specify hash method for Rfc2898DeriveBytes (#765)

piksel committed:
fix(tar): read full extended headers (#675)

* fix(tar): read full extended headers

* Update src/ICSharpCode.SharpZipLib/Tar/TarExtendedHeaderReader.cs

yihezkel committed:
fix: replace uses of obsolete method `RNGCryptoServiceProvider` (#766)

Replace insecure obsolete method (new RNGCryptoServiceProvider()) with RandomNumberGenerator.Create() in PkzipClassic, ZipFile and ZipOutputStream.

See docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rngcryptoserviceprovider?view=net-6.0 and dotnet/runtime#40169

sensslen committed:
fix(tar): enable unix paths in tar RootPath (#582)

Co-authored-by: nils måsén <nils@piksel.se>
Co-authored-by: Simon Ensslen <simon.ensslen@griesser.ch>

piksel committed:
fix(zip): handle iterating updated entries in ZipInputStream (#642)

piksel committed:
feat(gzip): add GzipOutputStream async support (#672)

piksel committed:
prepare for v1.4.0 (#755)

changes needed for the upcoming release, mainly related to .NET 6 compatibility

piksel committed:
docs: fix circular refs and target fw (#775)

piksel committed:
ci: build before generating docs

piksel committed:
ci: run docs generation on windows

piksel committed:
ci: use choco docfx on windows

piksel committed:
ci: fix typo in config path

piksel committed:
ci: split docs gen to build/deploy

piksel committed:
ci: fix config keys for deploy

remogloor committed:
Apply code refactoring suggested in review

remogloor committed:
Make it more explicit how big Large dummy data and small dummy data has to be.
